File: revision/missing_number.py
from typing import List
def missingNumber(nums: List[int]) -> int:
    '''
    [3, 0 , 1]
    partial_sum = 4 
    actual sum = n * (n + 1) / 2 
    i.e   3 * 4 / 2
    i.e   6 
    actual - partial_sum = answer
    '''
    #using sum formula 
    #return (len(nums) * (len(nums) + 1) // 2 ) - sum(nums)
    #using bit manupulation 
    xor1, xor2 = 0, 0
    for x in nums:
        xor1 ^= x
    for y in range(len(nums) + 1):
        xor2 ^= y
    return xor1 ^ xor2
nums = list(map(int, input().split()))
print(missingNumber(nums))

File: revision/missing_number.cpp
#include<iostream>
#include<vector>
using namespace std;
void input(vector<int>&nums){
   for(auto &i : nums){
       cin >> i;
   }
}
void display(vector<int>nums){
    for(auto const i : nums){
        cout << i << ' ' ;
    }
    cout << endl;
}
int missingNumber(vector<int>&nums){
    //using sum method 
    int partial_sum = 0;
    for(auto i : nums){
        partial_sum += i;
    }
    return (nums.size() * (nums.size() + 1) / 2) - partial_sum;
    //using bit manipulation 
    int xor1 = 0, xor2 = 0;
    for(auto i: nums){
        xor1 ^= i;
    }
    for(int i = 0; i < nums.size(); i++){
        xor2 ^= i;
    }
    return xor1 ^ xor2;
}
int main()
{
    int n;
    cin >> n;
    vector<int>nums(n);
    input(nums);
    cout << missingNumber(nums) << endl;
    return 0;
}